RFC-0020-Distributed-Pipeline-Parallelism.md

```diff
@@ -0,0 +1,209 @@
+
+1. Selected approach: "Approach 3 with Modifications"
+    * Inherit RemoteModule + torchpipe-based implementation from fairscale [experimental distributed pipeline parallelism](https://github.com/facebookresearch/fairscale/tree/main/fairscale/experimental/nn/distributed_pipeline).
+    * Switch autograd off of distributed autograd and onto manual handling of autograd in the pipeline, to facilitate implementing schedules (e.g. 1F1B)
+    * Abstract the runtime for each RemoteModule to allow for programming in execution schedules
+    * Switch from using DistributedLoss to having a loss callback to facilitate the last pipeline stage calling the loss locally rather than relying on the training loop to calculate the loss via RPC and call distributed autograd. This will be necessary with arbitrary schedules.
+
+### P0: Implement support for passing arbitrary data-types between pipeline stages
+
+Existing approaches that support this (in no particular order):
+
+* Some amount of [support](https://github.com/pytorch/pytorch/issues/53952) in existing PT implementation
+
+Proposed approach short-list:
+
+1. Hopefully should just work out of the box with the RPC API, but need to keep it in mind.
+
+### P0: 1.11 Prototype Release and out-of-tree demo on HF Transformers
+
+* Release API as prototype in the 1.11 release to facilitate gathering feedback
+* Validation: Out-of-tree demo on HF transformers repo - hack it together to get it to work and pull out work items to improve the API to remove places where code edits are needed
+* 1.11 Release Dates
+    * Feature submission: 11/30 EOD
+    * Branch cut 1/31/2022
+
+
+
+### **P0**: Support off-the-shelf schedules (1F1B or interleaving)
+
+Existing approaches that support this (in no particular order):
+
+* Megatron hardcoded schedules: [1f1b](https://github.com/NVIDIA/Megatron-LM/blob/5ac5571ba0265af4c491ee0af1508ca7589450c6/megatron/schedules.py#L517), [interleaved](https://github.com/NVIDIA/Megatron-LM/blob/5ac5571ba0265af4c491ee0af1508ca7589450c6/megatron/schedules.py#L187)
+
+Proposed approach short-list:
+
+1. "Approach 3 with Modifications"
+    * Once manual autograd is handled, and we abstract the workers, we can implement 1F1B or interleaved 1F1B using that infrastructure.
+
+### P0: Composability of PP with TP and DP (3d Parallelism)
+
+Existing approaches that support this (in no particular order):
+
+* `torch.distributed` APIs via module wrapper composition
+* [DeepSpeed](https://www.deepspeed.ai/tutorials/pipeline/)
+
+Proposed approach short-list:
+
+1. DistributedDataParallel wraps PipelineParallel API which operates on upcoming ShardedTensors
+2. Unified programming model (Stage 5)
+
+## Stage 3: Figure out how to reconcile local pipeline vs. distributed pipeline (2022H1)
+
+The existing approaches live in different corners of a 2-dimensional space with axes on **single-driver vs. actors** and **local vs. distributed**.
```

Comment:
What does `local` and `distributed` mean here?

Reply:
`local` means within a node, i.e. it's drawing a distinction between `torchgpipe`, which has a single driver process issuing commands onto CUDA streams v.s. something that operates across multiple nodes (and which necessarily will need to have multiple processes)